eip-0020.md

```diff
@@ -0,0 +1,202 @@
+Wallet also supports [EIP-0019](eip-0019.md) and private keys are not available,
+then the sign button behaves like a "Cold Sign" button according to EIP-0019. 
+
+7) The wallet application signs the transaction either using local private keys
+or using Cold Wallet and EIP-0019 protocol. The result of signing is
+`SignedTransaction` data.
+
+8) The Hot Wallet obtains the transaction id and sends it to the dApp using
+`ErgoPayTransactionSent` API post message, if successful the wallet then submits
+`SignedTransaction` to the blockchain.
+
+9) dApp monitors the transaction by id and proceeds with its business logic upon
+receiving enough confirmations. This concludes the ErgoPay protocol.
+
+Additional requirements: 
+- many instances of ErgoPay protocol MUST be able to run simultaneously withing
+the dApp
+- the Hot Wallet MAY implement "one-at-a-time" signing. When the signing
+is interrupted for some reason it can be started from the step 3) (i.e. scanning
+the QR)
+- when dApp receives the transaction id, but the `SignedTransaction` is not
+accepted to the blockchain then the dApp SHOULD detect this situation and stop
+monitoring that transaction.
+
+## Data Formats
+
+The data formats of this EIP are based on a new binary data structure and
+serialization format called `ReducedTransaction` which is described in
+[EIP-0019](eip-0019.md).
+
+Below we describe the data formats specified by this EIP and refer to the
+formats defined in EIP-0019.
+
+Depending on the bytes size of `ReducedTransaction` it may not fit into 4K limit
+of QR codes. In this case or for whatever reason the dApp may encode the
+following data in the QR code on the payment screen or link to it:
+
+`ergopayurl://<URL without https>`, for example
+`ergopayurl://sigmausd.io/signingRequest/2001-16b8-66c4-b800-6e52-8ce4`
+
+When the Wallet application parses the QR code bytes it can use prefix URI scheme
+to learn the message type.
+
+The wallet application should request URL and obtain the following data (json format)
+
+```
+ErgoPaySigningRequest:
+  - transaction: ReducedTransaction
+  - description: String (optional)
+  - replyToUrl: String (optional)
+```
+
+When dapp does not make use of an extra request, the `ErgoPaySigningRequest` could 
+also be encoded in the QR code or link:
+
+`ergopay:<ReducedTransaction, base 64 url safe encoded>`
+
+Once the transaction details obtained, they CAN BE OPTIONALLY shown on the payment
+screen of the Hot Wallet application.
+
+After signing is performed and the `SignedTransaction` data is obtained, the 
+wallet app need to POST the following data to the dApp using `replyToUrl` from the
+signing request (json format).
+
+```
+ErgoPayTransactionId:
+  - txId: String
+```
+
+After `ErgoPayTransactionId` is sent, only then the dApp can submit the
+SignedTransaction to the blockchain. The order is important, otherwise, if
+transaction is included into blockchain and dApp notification fails the dApp
```

Comment:
I don't understand the argument here.
First of all, not all dapps can support `replyToUrl` because they may simply be serverless (like the current AH, ErgoUtils and SigmaUSD) so they can not provide an API for that.
I think a more general solution is needed here. In the simplest case, the dapp can keep monitoring for the outputs that it expects to see if they are in the mempool.

Reply:
Also, if I remember correctly, an unsigned tx ID can be obtained. So I think the dapp can know the tx ID even before it being signed. @kushti can clarify this.

Reply:
@anon-real yes, indeed unsignedTx.id == signedTx.id, so this requirement can be removed.
I clarified this point in the text.